### PR TITLE
feat: implement file-based prompt arguments with YAML frontmatter (#3049)

### LINKED ISSUE #3049: Feature request: File-based prompts lack structured argument support
### Checks

- [x] I have searched [github.com/aws/amazon-q-developer-cli/issues](https://github.com/aws/amazon-q-developer-cli/issues?q=) and there are no duplicates of my issue
- [x] I have run `q doctor` in the affected terminal session
- [x] I have run `q restart` and replicated the issue again

### Operating system

macOS 15.7.0 (24G222)

### Expected behaviour

<This will be visible to anyone. Do not include personal or sensitive information>

File-based prompts should support structured arguments like MCP prompts, including argument validation, required/optional distinctions, and metadata display in /prompts list.

### Actual behaviour

<This will be visible to anyone. Do not include personal or sensitive information>

Here is my file-based prompt

``` txt
~/.aws/amazonq/prompts/GetDateYYYYMMDD.md

# Get date

Show only date in {time}
format: YYYYMMDD

Don't show anything except date
```

Here is result when run `/prompt list` comparing file-based prompt vs prompt from mcp server with the same prompt:

``` txt
Prompt                                    Description                              Arguments (* = required)
▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔
Global (.aws/amazonq/prompts):
- GetDateYYYYMMDD

prompt-server (MCP):
- MCP-GetDateYYYYMMDD    Get date                                 time*
```

When run `@GetDateYYYYMMDD yesterday`

```
> @GetDateYYYYMMDD yesterday

# Get date

Show only date in {time}
format: YYYYMMDD

Don't show anything except date

🛠️  Using tool: execute_bash
 ⋮
 ● I will run the following shell command: date +%Y%m%d
 ⋮
 ↳ Purpose: Get current date in YYYYMMDD format


Allow this action? Use 't' to trust (always allow) this tool for the session. [y/n/t]:

[vanilla] > y

20250930

 ⋮
 ● Completed in 0.33s

> 20250930
```

When run `@MCP-GetDateYYYYMMDD yesterday`

```
> @MCP-GetDateYYYYMMDD yesterday

# Get date

Show only date in yesterday
format: YYYYMMDD

Don't show anything except date

🛠️  Using tool: execute_bash
 ⋮
 ● I will run the following shell command: date -v-1d +%Y%m%d
 ⋮
 ↳ Purpose: Get yesterday's date in YYYYMMDD format

Allow this action? Use 't' to trust (always allow) this tool for the session. [y/n/t]:

[vanilla] > y

20250929

 ⋮
 ● Completed in 0.43s

> 20250929
```

### Steps to reproduce

<This will be visible to anyone. Do not include personal or sensitive information>

1. Create file-based prompt with {language} and {name} placeholders
2. Use prompt with missing arguments
3. No validation occurs
4. /prompts list shows no argument information

### Environment

```yaml
<This will be visible to anyone. Do not include personal or sensitive information>

[q-details]
version = "1.16.2"
hash = "c4c80bb44cb22b6e19129ed7475f2793ff20135d"
date = "2025-09-19T01:27:56.356289Z (11d ago)"

[system-info]
os = "macOS 15.7.0 (24G222)"
chip = "Apple M3 Pro"
total-cores = 11
memory = "18.00 GB"

[environment]
cwd = "/Users/USER/Code/amazon-q-developer/references/amazon-q-developer-cli"
cli-path = "/Users/USER/Code/amazon-q-developer/references/amazon-q-developer-cli"
install-method = "unknown"

[env-vars]
PATH = "/Users/USER/.nvm/versions/node/v22.15.1/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Applications/iTerm.app/Contents/Resources/utilities:/Users/USER/.local/bin:/Users/USER/.local/bin:/Applications/Visual Studio Code.app/Contents/Resources/app/bin:/Users/USER/.local/bin:/Users/USER/.toolbox/bin"
QTERM_SESSION_ID = "4e61b01d6f784ad7a57e6a6a070e7d6e"
Q_SET_PARENT_CHECK = "1"
Q_TERM = "1.16.2"
SHELL = "/bin/zsh"
TERM = "xterm-256color"
__CFBundleIdentifier = "com.googlecode.iterm2"

[chat-settings]


[chat-trusted_tools]
@calculator=trusted
@awslabs.aws-pricing-mcp-server=trusted
@fetch=trusted
fs_read=trusted
@awslabs.aws-api-mcp-server/suggest_aws_commands=trusted
@awslabs.aws-documentation-mcp-server=trusted
execute_bash=trusted
@aws-knowledge-mcp-server=trusted
@awslabs.core-mcp-server=trusted

[chat-failed_request_ids]
none

[chat-context]
current_profile=default
profile_context=
/Users/sawitmee/Library/CloudStorage/OneDrive-amazon.com/sawitmee/Q-Context/based.md

files=
/Users/sawitmee/Library/CloudStorage/OneDrive-amazon.com/sawitmee/Q-Context/based.md, 70 tkns
total context size=70 tkns
```

### PR DESCRIPTION
**Pull request by peepeepopapapeepeepo**
## Summary
Implements optional YAML frontmatter and automatic argument detection for file-based prompts as requested in #3049.

## Changes
- **YAML Frontmatter Parsing**: Optional headers with description and arguments
- **Auto-Detection**: Automatically detects `{{placeholder}}` patterns in prompt content
- **Enhanced Display**: `/prompts` command now shows arguments even without descriptions
- **Argument Processing**: Validates and replaces placeholders with provided values
- **Comprehensive Testing**: 24+ tests covering all scenarios and edge cases
- **Documentation**: Complete user guide with examples

### 1. Simple Auto-Detection
```markdown
Get weather for {{city}} in {{units}} format.
```

### 2. Description + Auto-Detection  
```markdown
---
description: "Get current date"
---
Today's date in {{format}} format using {{timezone}}.
```

### 3. Full YAML Control
```markdown
---
description: "Get weather information"
arguments:
  - name: "city"
    required: true
  - name: "units"
    required: false
---
Get weather for {{city}} in {{units}} format.
```

## Backward Compatibility
✅ All existing prompts continue to work unchanged
✅ No breaking changes to existing functionality
✅ Optional feature activated only when YAML or `{{}}` patterns detected

## Testing
- All existing tests pass (24/24)
- New comprehensive test coverage for parsing, validation, and edge cases
- Manual testing verified with real prompt files

## Code Quality
- Follows existing code style and patterns
- Functions organized into logical sections
- Proper error handling and input validation
- No clippy warnings, clean compilation

Closes #3049

### CODE REVIEW

**Comment by tobinquadros** [comment]
Looks useful. I have also found that the arguments in many of my prompts are pre-defined choices, that would be a useful addition along with the `name` and `required` values.

**Comment by peepeepopapapeepeepo** [comment]
> Looks useful. I have also found that the arguments in many of my prompts are pre-defined choices, that would be a useful addition along with the `name` and `required` values.

@tobinquadros That's a useful suggestion! Could you provide an example of how you'd like to see the pre-defined choices work? For instance:

``` markdown
---
description: "Get weather information"
arguments:
  - name: "city"
    required: true
    choices:
      - Bangkok
      - Tokyo 
  - name: "units"
    required: false
    choices:
      - celsius
      - fahrenheit 
    default: celsius
---
```

**Comment by shafkevi** [comment]
This would be a great addition! Hopefully it can get added in.

**Comment by eoliphan** [comment]
Q needs this (or something similar) ASAP.  It's table stakes. Parameterized, "Higher order", etc. prompts are typically the very first step to standardizing workflows.  Not having this also is a blocker for using q cli with other frameworks  like bmad, spec kit, etc.

File: crates/chat-cli/src/cli/chat/cli/prompts.rs [hunk 1]
```diff
@@ -78,6 +88,179 @@ pub enum GetPromptError {
...
+            return Err(PromptArgumentError::InvalidYaml(
+                "Frontmatter missing closing '---'".to_string(),
+            ));
+        }
+    } else {
+        (content.to_string(), None)
+    };
+
+    // Auto-detect arguments from {{}} placeholders in the body
+    let mut detected_args = std::collections::HashSet::new();
```
Comment (kkashilk) [inline crates/chat-cli/src/cli/chat/cli/prompts.rs:155]: If we wanna provide a consistent arguments order, we can use a Vec to store the args, and the HashSet to check if that arg has already been added before.

File: crates/chat-cli/Cargo.toml [hunk 2]
```diff
@@ -86,6 +86,7 @@ semantic_search_client.workspace = true
 semver.workspace = true
 serde.workspace = true
 serde_json.workspace = true
+serde_yaml = { version = "0.9", default-features = false }
```
Comment (kkashilk) [inline crates/chat-cli/Cargo.toml:89]: Add this here - https://github.com/aws/amazon-q-developer-cli/blob/b3fa2264bb2010d7c7e6f3cbe7be67a5c5b3c168/Cargo.toml and then just enable it in this Cargo.toml file

File: docs/prompts.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+File-based prompts allow you to create reusable templates with dynamic arguments. They are stored as `.md` files in your prompts directories.
+
+## Prompt Locations
+
+- **Global**: `~/.aws/amazonq/prompts/`
+- **Local**: `.amazonq/prompts/` (project-specific)
+
+## Creating Prompts
+
+### Simple Prompts with Auto-Detection
```
Comment (kkashilk) [inline docs/prompts.md:12]: nit: `Simple Prompts with Auto-Detected Arguments`

File: docs/prompts.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+The easiest way to create a prompt is to simply use `{{placeholder}}` syntax in your markdown file:
```
Comment (kkashilk) [inline docs/prompts.md:14]: nit: `The easiest way to create a prompt with arguments is ..`

File: docs/prompts.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+Get weather for {{city}} in {{units}} format.
+```
+
+**Usage:** `@weather Tokyo celsius`
+- First argument (`Tokyo`) replaces `{{city}}`
+- Second argument (`celsius`) replaces `{{units}}`
+
+### Argument Order
+
+**For auto-detected arguments:** Order is determined by first appearance in the text.
```
Comment (kkashilk) [inline docs/prompts.md:99]: that's not the case currently, see this example from Description - 

```
/prompts details simple-weather


Prompt Details
▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔

Name: simple-weather
Source: /home/sawitmee/.aws/amazonq/prompts/simple-weather.md

Usage: @simple-weather [units] [city]

Arguments:
  (optional) units
  (optional) city

Usage: @simple-weather

Content Preview:
  Get weather for {{city}} in {{units}} format.
```

File: crates/chat-cli/src/cli/chat/cli/prompts.rs [hunk 6]
```diff
@@ -1325,16 +1696,28 @@ impl PromptsSubcommand {
                 execute!(session.stderr)?;
             }
 
+            // Process arguments if metadata and arguments exist
+            let final_content = if let (Some(meta), Some(args)) = (&metadata, &arguments) {
+                if let Some(prompt_args) = &meta.arguments {
+                    process_file_arguments(&content, prompt_args, args)
```
Comment (kkashilk) [inline crates/chat-cli/src/cli/chat/cli/prompts.rs:1702]: We can do prompt-name validation `parse_frontmatter_if_exists` so that we surface any errors early on

File: crates/chat-cli/src/cli/chat/cli/prompts.rs [hunk 7]
```diff
@@ -78,6 +88,179 @@ pub enum GetPromptError {
...
+        if let Some(end_pos) = stripped.find("\n---\n") {
+            let yaml_content = &stripped[..end_pos];
+            let body = &stripped[end_pos + 5..]; // +5 for "\n---\n"
+
+            let metadata: FilePromptMetadata =
+                serde_yaml::from_str(yaml_content).map_err(|e| PromptArgumentError::InvalidYaml(e.to_string()))?;
+
+            // Basic validation
+            if let Some(args) = &metadata.arguments {
+                for arg in args {
```
Comment (kkashilk) [inline crates/chat-cli/src/cli/chat/cli/prompts.rs:135]: Inherent assumption here is that you can either have frontmatter args or placeholder args, but not both.

Let's either clarify that in the `prompts.md` file, or we can consider appending placeholder args that are not in the frontmatter as optional args.
Thoughts?

File: crates/chat-cli/src/cli/chat/cli/prompts.rs [hunk 8]
```diff
@@ -833,6 +1002,113 @@ impl PromptsArgs {
...
+                UnicodeWidthStr::width(" (overrides global)")
+            } else {
+                0
+            };
+            let current_pos = name_width + override_width;
+
+            if let Some(ref meta) = metadata {
+                let desc_width = if let Some(ref desc) = meta.description {
+                    let truncated_desc = truncate_description(desc, 40);
+                    let description_padding = description_pos.saturating_sub(current_pos);
```
Comment (kkashilk) [inline crates/chat-cli/src/cli/chat/cli/prompts.rs:1066]: what happens when `prompt_name (overrides global)` is longer than the calculated `description_pos` value above?

File: crates/chat-cli/src/cli/chat/cli/prompts.rs
Comment (kkashilk) [inline crates/chat-cli/src/cli/chat/cli/prompts.rs:1481]: Can we not combine the `display_file_prompt_details` and `display_prompt_details` methods? Seems like a lot of duplication

File: docs/prompts.md (hunk 5 above)
Comment (kkashilk) [reply docs/prompts.md:99]: We'll need to use a Vec to parse the arguments, as mentioned in my comment in prompts.rs file above ^

**Comment by richzhu369** [comment]
can use [spec kit](https://github.com/github/spec-kit) now ?
